### PR TITLE
Persist user ID to replay data

### LINKED ISSUE #24624: Local scores shows other person profile picture and flags
### Type

Cosmetic

### Bug description

After importing a offline score from stable that has, let's say, "mrekk" as its player name, when logging into Lazer and seeing the local leaderboard, the game shows mrekk profile picture and flag as he was the one that originally set the score, even tho it was a offline score.

I noticed that after importing my scores using the setup wizard, that's when I noticed the local score confused a local score of mine in this same map with the name "gigi", with the player GIGI. The same happened when I imported this single "mrekk" replay into lazer.

image

### Screenshots or videos

_No response_

### Version

2023.815.0-lazer

### Logs

[performance.log](https://github.com/ppy/osu/files/12413101/performance.log)
[runtime.log](https://github.com/ppy/osu/files/12413102/runtime.log)
[updater.log](https://github.com/ppy/osu/files/12413103/updater.log)
[database.log](https://github.com/ppy/osu/files/12413104/database.log)
[input.log](https://github.com/ppy/osu/files/12413105/input.log)
[network.log](https://github.com/ppy/osu/files/12413106/network.log)

### PR DESCRIPTION
**Pull request by bdach**
RFC. As proposed in https://github.com/ppy/osu/issues/24624#issuecomment-1918401500. Would fix https://github.com/ppy/osu/issues/24624 going forward.

Please give the `ScoreImporter` changes a read first and foremost as I foresee them being the most dubious of the bunch. Rationale as to why they are done the way they are is in the inline comment there.

- [x] Needs to be tested against server spectator as I'm not sure whether this will do the correct thing there (will do this myself whenever able)

### CODE REVIEW

**Comment by bdach** [comment]
This has been tested full-stack now. While https://github.com/ppy/osu-server-spectator/pull/230 is not a strict dependency of this I'd recommend to get it in first for the peace of mind at least if this is going to go in.

**Comment by smoogipoo** [comment]
Tests are failing here

File: osu.Game/Scoring/ScoreImporter.cs [hunk 1]
```diff
@@ -127,15 +135,54 @@ private void populateUserDetails(ScoreInfo model)
             if (model.RealmUser.OnlineID == APIUser.SYSTEM_USER_ID)
                 return;
 
+            if (model.RealmUser.OnlineID > 1)
+            {
+                model.User = lookupUserById(model.RealmUser.OnlineID) ?? model.User;
```
Comment (smoogipoo) [inline osu.Game/Scoring/ScoreImporter.cs:140]: This is not going to scale. We're doing a full user lookup with every score import? What about when someone wants to import all their scores from stable?

I think this needs to be a post process.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Scoring/ScoreImporter.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Scoring/ScoreImporter.cs:140]: I'm not sure why this is a review concern in the context of this pull given

- we're already doing this on `master`
- there is local caching in place to specifically address the "import from stable" concern
- the other existing user cache cannot be used due to reasons explained in the inline commentary added in this pull

Comment (smoogipoo) [reply osu.Game/Scoring/ScoreImporter.cs:140]: I didn't notice the implementation of `lookupUserByName`.

**Comment by bdach** [comment]
> Tests are failing here

Should be better now.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**